### ISSUE TITLE
"Update automatic entity labels" bulk action not available to views

### ISSUE THREAD

**Issue by argiepiano** [backdrop-contrib/auto_entitylabel#8]
Auto Entity Label defines an action, `auto_entitylabel_entity_update_action` that is not available to core's Bulk Operations in Views. 

The problem is that `auto_entitylabel_action_info()` defines `'type' => 'entity'`  for this action. Unlike D7's VBO, Backdrop core Bulk Operations do not accept  `entity` as a type for actions. See documentation [here](https://github.com/backdrop/backdrop/blob/52138cb8eca87a9fcb75c6536c5517d53fc6a7af/core/modules/system/system.api.php#L3429
). 

Instead, the module should create individual, separate actions for each core entity (unfortunately). This is a Backdrop core limitation.

**Comment by argiepiano** [comment]
An alternative is to modify core to accept type `entity` for core's Bulk Operations.

**Comment by yorkshire-pudding** [comment]
@argiepiano - out of interest, does the [Views Bulk Operation contrib module](backdrop-contrib/views_bulk_operations) work with this module, thereby overcoming this limitation?

**Comment by argiepiano** [comment]
Yes. But the hook used to define actions for VBO has changed. It's `hook_vboaction_info()`. Check out the readme of VBO.  You can basically rename the function `auto_entitylabel_action_info()` to `auto_entitylabel_vboaction_info()` and add the `'vbo' => TRUE` key to it. It should work without further modifications.

**Comment by yorkshire-pudding** [comment]
Mmm - without any modification I used this module together with VBO and it worked. The entity is a node so is this specifically about custom entities?

**Comment by argiepiano** [comment]
Oh, yes, that's true, but the action is considered a "core" action since it's defined by hook_action_info(), instead of hook_vboaction_info().

As long as the action doesn't have any configuration form before running, then it should work fine. The only drawback is that core actions do not run in batch calls by default. So, if you have tons of entities to update, it's best to make the change to hook_vboaction_info as I described above, so that VBO's batched processing is used.

**Comment by herbdool** [comment]
@argiepiano I welcome a PR here. Maybe we change it to use `hook_vboaction_info()` since it's not working for core anyway.

**Comment by herbdool** [comment]
@argiepiano can you take a look at my PR? It checks which entity types have automatic labels enabled and then provides actions for those. Each action uses the same callback so we don't have to duplicate code. Maybe we could add `hook_vboaction_info()` as well so that it can use vbo batching if using VBO.

**Comment by herbdool** [comment]
I decided to have both hooks. I think this will make it backwards-compatible if using the action with VBO already. It means there's a bit of duplication if VBO is enabled, but I think that's acceptable:

<img width="480" height="248" alt="Image" src="https://github.com/user-attachments/assets/8e5cc2b1-871d-4c8e-911f-911a205c722a" />

**Comment by argiepiano** [comment]
I'm running into many warnings when using this module and patch in a site with many custom entities. 

```
Warning: Trying to access array offset on null in auto_entitylabel_action_info() (line 364 of /Users/XXX/Sites/localhost/XXX/modules/auto_entitylabel-1.x-1.x/auto_entitylabel.module).
```

The above happens to entities that don't use the Field API admin (mostly non-fieldable entities, or the so-called "configuration entities" that Entity Plus provides, which are legacy entities used to define bundles of fieldable entities).

We need some defensive coding to avoid looking up `$label_settings['status'])` for entities that are not listed in `auto_entitylabel.settings`.  

You can see this problem in action by installing `registration`, which uses one of those so-called "configuration entities".

**Comment by herbdool** [comment]
@argiepiano I added a fix. Other than that are you okay with the approach?

**Comment by argiepiano** [comment]
I didn't do full testing after I encountered the warnings. Let me go back to it later today.

**Comment by argiepiano** [comment]
Now I'm getting 

`Warning: Undefined variable $info in auto_entitylabel_action_info() (line 380 of /Users/XXX/Sites/localhost/XXX/modules/auto_entitylabel-1.x-1.x/auto_entitylabel.module).`

This is because $info is not declare after your change.

May I suggest you install `registration` and test with it?

**Comment by herbdool** [comment]
@argiepiano thanks for catching that. Fixed.

I had installed registration; that wasn't the problem. You probably saw the error because you had no entities with auto_entitylabel enabled?

**Comment by argiepiano** [comment]
I'll get back to this tomorrow

**Comment by argiepiano** [comment]
@herbdool while the approach seems fine, I'm finding other issues with custom entities. For example the machine name of the column that contains the label for custom entities is assumed to be `title`, but this is not always the case. This is hard-coded in `_auto_entitylabel_get_label()`. I'm testing with a custom entity called `teacher` where the label of the entity is saved in a column (field) called `name`. Therefore, while I get no error, the module is useless for this case.

This is clearly a problem that goes beyond this issue (bulk actions seem to work for core entities). But if we are going to enable VBO with type `entity`, we'd better make sure it works...

**Comment by herbdool** [comment]
This seems to be a pre-existing problem where a tab would appear even if there is no label field, by providing a default which is wrong. So I've changed it to pull up the actual label from `entity_get_info()`. If there are other problems we can address them on new issues. I'm merging this one.

**Comment by argiepiano** [comment]
@herbdool I'm not sure you understood my comment. This module doesn't work with custom entities that use a column other than `title` for their labels. As I pointed out, `title` is hard-coded  by function  `_auto_entitylabel_get_label()` for all entities that are not in core.

So even though the VBO integration seems to work for core entities, it does not work at all for those types of custom entities.

I think it's unfortunate you went ahead and did a release despite this bug, despite the fact that I took the time to review it and pointed it out. Anyway, I don't use the module, so I guess it will be up to the next bystander who finds the module doesn't work as expected.

**Comment by herbdool** [comment]
@argiepiano I did understand. I had changed the function to:

```
function _auto_entitylabel_get_label($entity_type) {
  $entity_info = entity_get_info($entity_type);
  if (isset($entity_info['entity keys']['label'])) {
    return $entity_info['entity keys']['label'];
  }
  return '';
}
```

Previously entity types like registration_type or flag had an Auto Label tab and now they do not. So in my opinion it's at least better than before: few entity types are getting Auto Label tab where it wasn't working before.

And sure, the action still appears for Views but it won't do anything if someone adds it and then attempts to update a label of some entity which doesn't actually have auto entity label settings. I think that's probably what was happening in Drupal 7 as well so no worse there either.

I wanted to get something which was no worse than before so I can move on to other work.

**Comment by argiepiano** [comment]
Oh, I see! Sorry, I had missed that last change. Looks good!
